### PR TITLE
[xc_admin] add update product metadata

### PR DESCRIPTION
**Pull request by cctdaniel**

### CODE REVIEW

File: governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    }
+  }, [anchorWallet, connection, connected, cluster])
+
+  return (
+    <div className="relative">
+      <Modal
+        isModalOpen={isModalOpen}
+        setIsModalOpen={setIsModalOpen}
+        closeModal={closeModal}
+        content={<ModalContent changes={productMetadataChanges} />}
```
Comment (guibescos) [inline governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:302]: Modal doesn't show which Symbol you are modifying

File: governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    // check if json keys are existing products
+    const jsonParsed = JSON.parse(json)
+    const jsonSymbols = Object.keys(jsonParsed)
+    const existingSymbols = Object.keys(data)
+    // check that jsonSymbols is equal to existingSymbols no matter the order
+    if (
+      JSON.stringify(jsonSymbols.sort()) !==
+      JSON.stringify(existingSymbols.sort())
+    ) {
+      toast.error('Symbols in json file do not match existing symbols!')
```
Comment (guibescos) [inline governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:160]: Can you error if there are duplicate keys in the uploaded JSON

File: governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            sortedInnerData[innerKey] = data[key][innerKey]
+          })
+        sortedSymbolToProductMetadataMapping[key] = sortedInnerData
+      })
+
+    return sortedSymbolToProductMetadataMapping
+  }
+
+  // function to download json file
+  const handleDownloadJsonButtonClick = () => {
```
Comment (guibescos) [inline governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:95]: Two fields in the downloaded json are not really editable price_account and symbol.
I suggest you don't show price_account (low-level info).
Symbol is already there as it is the key, so you can get away with not showing.

If someone tries to edit/add these fields, raise a toast error.

File: governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    let isValid = true
+    // check that the keys of the values of json are equal to the keys of the values of data
+    jsonSymbols.forEach((symbol) => {
+      const jsonKeys = Object.keys(jsonParsed[symbol])
+      const existingKeys = Object.keys(data[symbol])
+      if (
+        JSON.stringify(jsonKeys.sort()) !== JSON.stringify(existingKeys.sort())
+      ) {
+        toast.error(
+          `Keys in json file do not match existing keys for symbol ${symbol}!`
```
Comment (guibescos) [inline governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:173]: I think it's fine if you want to add new keys to the metadata/ delete previous keys.

**Review by guibescos — APPROVED** [review]
Thank you sir, fix comments please

File: governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx (hunk 2 above)
Comment (cctdaniel) [reply governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:160]: this is a general error that includes this situation -- do you mean do a specific check just for duplicate keys?

Comment (cctdaniel) [reply governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:160]: do you mean json keys in the metadata?

Comment (guibescos) [reply governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:160]: Duplicate symbol

File: governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx (hunk 1 above)
Comment (cctdaniel) [reply governance/xc_admin/packages/xc_admin_frontend/components/tabs/UpdateProductMetadata.tsx:302]: good catch

**Pull request merged**